### ISSUE TITLE
Unable to make HTTPS requests with C#'s built-in class (not HTTPRequest) (fixed in `master`)

### ISSUE THREAD

**Issue by esbennn** [godotengine/godot#36958]
**Godot version:** 3.2 stable 

**OS/device including version:** Windows 10

**Issue description:**
We have a C# library, that includes a Log service, which we use in most our projects (of any kind). This service runs in a thread an POST's a bunch of strings to our server. 
Implementing this Log service works just fine on my Linux machine, but when i deploy on a Windows machine (or my coworker runs it in the editor on his Windows machine), the web request fails with 
`Error: SecureChannelFailure (The authentication or decryption has failed.)`.
Here's a full stack trace:
```
at System.Net.WebConnection.CreateStream (System.Net.WebOperation operation, System.Boolean reused, System.Threading.CancellationToken cancellationToken) [0x0021a] in <8c9baba7add14c22be562e4c18e5738f>:0
  at System.Net.WebConnection.InitConnection (System.Net.WebOperation operation, System.Threading.CancellationToken cancellationToken) [0x00141] in <8c9baba7add14c22be562e4c18e5738f>:0
  at System.Net.WebOperation.Run () [0x0009a] in <8c9baba7add14c22be562e4c18e5738f>:0
  at System.Net.WebCompletionSource`1[T].WaitForCompletion () [0x00094] in <8c9baba7add14c22be562e4c18e5738f>:0
  at System.Net.HttpWebRequest.GetRequestStream () [0x00016] in <8c9baba7add14c22be562e4c18e5738f>:0
  at Playground.Sandbox.Utility.WebRequest.Post (System.String url, System.Collections.Generic.Dictionary`2[TKey,TValue] values) [0x000b6] in <efaba92892e84c979ad8940e61632f93>:0
  at Playground.Sandbox.Services.LogService`1[T].Upload () [0x001f8] in <efaba92892e84c979ad8940e61632f93>:0 : Error: SecureChannelFailure (The authentication or decryption has failed.)
```
If i remove the *s* from the url, and stop our server from redirecting to https, it works, so this  definitely has to do with tls. 

I have tried overriding the ServerCertificateValidationCallback with 
```
ServicePointManager.ServerCertificateValidationCallback += (p1, p2, p3, p4) => true;
``` 
as per [this stackoverflow post](https://stackoverflow.com/questions/4926676/mono-https-webrequest-fails-with-the-authentication-or-decryption-has-failed/49799509#49799509)

I have also tried forcing tls 1.2: 
```
System.Net.ServicePointManager.SecurityProtocol = System.Net.SecurityProtocolType.Tls12;
```

According to [monos documentation, you can import certificates with a tool called *mozroots*, but it does not seem to be bundled with Godot.](https://www.mono-project.com/docs/faq/security/) Is there a way to use this tool from my code, so I don't have to manually run it for every pc i deploy to?

**Steps to reproduce:**
Make a https request from C#/mono on a Windows machine

**Minimal reproduction project:**
New C# project, that creates an HTTP request to a https-url, using either System.Net.WebRequest or System.Net.WebClient.

**Comment by nonunknown** [comment]
Did you tried to recreate the issue using gdscript, if using gdscript the problem may be with c# lib, not the engine itself!

**Comment by groud** [comment]
Sorry, this place is to report bug about Godot, not really about other libraries.

Regarding your question about how you could use mozroot inside Godot, you should ask on the other community channels first.

**Comment by esbennn** [comment]
> Sorry, this place is to report bug about Godot, not really about thirdparty libraries.
> 
> Regarding your question about how you could use mozroot inside Godot, you should ask on the other community channels first.

Well, the point is that making http requests as one normally would in C# doesn't seem to work with how Mono is implemented - whether this is in a 3rd party library or not

**Comment by neikeq** [comment]
Yes, this is not a bug in a thitdparty library, but Godot and how it uses mono.

**Comment by groud** [comment]
> Yes, this is not a bug in a thitdparty library, but Godot and how it uses mono.

I mean, if OP does not use a Godot internal class (like HTTPRequest), how can this be a Godot issue ? We don't have the hand on the code made by the user.

**Comment by neikeq** [comment]
What happens if you run Godot with the environment variable `MONO_TLS_PROVIDER=btls`?

**Comment by neikeq** [comment]
Also, for clarification, on Windows this works on the Godot editor but not exported games?

**Comment by esbennn** [comment]
> What happens if you run Godot with the environment variable `MONO_TLS_PROVIDER=btls`?
I tried putting `Environment.SetEnvironmentVariable("MONO_TLS_PROVIDER", "btls");` in the _EnterTree function of my root node. 
That gives this exception:
```
at System.Net.WebConnection.CreateStream (System.Net.WebOperation operation, System.Boolean reused, System.Threading.CancellationToken cancellationToken) [0x00207] in <8c9baba7add14c22be562e4c18e5738f>:0
  at System.Net.WebConnection.InitConnection (System.Net.WebOperation operation, System.Threading.CancellationToken cancellationToken) [0x00141] in <8c9baba7add14c22be562e4c18e5738f>:0
  at System.Net.WebOperation.Run () [0x0009a] in <8c9baba7add14c22be562e4c18e5738f>:0
  at System.Net.WebCompletionSource`1[T].WaitForCompletion () [0x00094] in <8c9baba7add14c22be562e4c18e5738f>:0
  at System.Net.HttpWebRequest.GetRequestStream () [0x00016] in <8c9baba7add14c22be562e4c18e5738f>:0
  at Playground.Sandbox.Utility.WebRequest.Post (System.String url, System.Collections.Generic.Dictionary`2[TKey,TValue] values) [0x000b6] in <7c62243f7d9f4a399e5f28f251f59a42>:0
  at Playground.Sandbox.Services.LogService`1[T].Upload () [0x001f8] in <7c62243f7d9f4a399e5f28f251f59a42>:0 : Error: ConnectFailure (TLS Support not available.)
```

**Comment by esbennn** [comment]
> Also, for clarification, on Windows this works on the Godot editor but not exported games?

No, it fails both in the editor and when exported.
Works on Linux though

**Comment by esbennn** [comment]
Followup:
Making a _very_ simple request like the following works:
```
WebClient client = new System.Net.WebClient();
string response = client.DownloadString("http://isitfriday.org/");
```
If the url is changed to `https...`, it does not work, but throws the Exception in the comment above.

I also tried using the RestSharp library to perform the requests - same story, works with http.
Using https the status code on the response is `0`, seemingly indicating that the request does not make it to the server.

**Comment by neikeq** [comment]
Thanks for the example code. I'll be testing this later this week.

**Comment by X52p** [comment]
Any news or workarounds for this so far? My best guess is, that it doesn't find the certificates.

By the way, I tried all of these:
<https://stackoverflow.com/questions/4926676/mono-https-webrequest-fails-with-the-authentication-or-decryption-has-failed>
non worked.

**Comment by Calinou** [comment]
@X52p You should be able to use Godot's HTTPRequest class from C#.

**Comment by esbennn** [comment]
> Any news or workarounds for this so far? My best guess is, that it doesn't find the certificates.
> 
> By the way, I tried all of these:
> https://stackoverflow.com/questions/4926676/mono-https-webrequest-fails-with-the-authentication-or-decryption-has-failed
> non worked.

I haver not been able to find a proper workaround, other than using a non-https endpoint - but i haven't tried it again since 3.2.2 came out.
As @Calinou says, the built-in HTTPRequest does work from C# with https, but dotnet standard approach (useful for threaded background tasks and such) does not

**Comment by X52p** [comment]
The problem is, that I'm using a library that is doing the requests. So I can't just replace the commands.
(currently using v3.2.2.stable.mono.official)

**Comment by esbennn** [comment]
> I'm using a library that is doing the requests.

Excactly

**Comment by X52p** [comment]
Did anyone try to use a different build tool? (In Editor Settings -> Mono -> Builds and than something other than VS Build Tools)
I tried but I could not get them to compile (It is not finding the libraries).

Is there any documentation on how to use them? Then I could set them up correctly and give it a try. (I'm new to using Mono and NuGet, sorry if this is a stupid question)

**Comment by X52p** [comment]
I just compiled Godot 3.2 with Mono myself. Also doing this before building:

`cert-sync --user myCert.pem`

Were myCert.pem is the certificate of the server I'm trying to connect to.
This seems to fix it. (I got other errors now, but these are probably library related)

**Comment by TechnoPorg** [comment]
I think this issue is because of how the official Godot binaries are built. In Mono, there's an `abstract` class called `MobileTlsProvider` used by the C# built-in web requests. A `MobileTlsProvider` is obtained via `MonoTlsProviderFactory`, which provides an implementation of `MobileTlsProvider` that will work with the system it's running on. It seems to me like for Windows, Mono uses the BTLS system and without it can't use TLS. BTLS is only available in Mono if `MONO_FEATURE_BTLS` is defined when [MonoTlsProviderFactory.cs](https://github.com/mono/mono/blob/main/mcs/class/System/Mono.Net.Security/MonoTlsProviderFactory.cs) is compiled.
Example:
```csharp
#if MONO_FEATURE_BTLS
		[MethodImpl (MethodImplOptions.InternalCall)]
		internal extern static bool IsBtlsSupported ();
#endif
```
The above function points to a C function that returns true if `--enable-btls` is passed to `autogen.sh`.
In my Mono 4.5 installation [from the binary on their website](https://www.mono-project.com/download/stable/), the BTLS code is still present when I look at it with DnSpy. In the Mono 4.5 from the [Godot 3.3 official binary](https://godotengine.org/download), however, the same code didn't get included. C# HTTPS requests work when you build Godot with Mono yourself, I think because it copies the existing Mono installation which for most people comes from the official Mono binaries.

My Godot build
![image](https://user-images.githubusercontent.com/69441745/127072539-e9866288-c059-4176-ad86-828f44baab61.png)

No BTLS in the official binary 
![image](https://user-images.githubusercontent.com/69441745/127072663-41c5b3b2-e2ac-41b7-9085-142c4db01af9.png)

**Comment by akien-mga** [comment]
Here's a build of Godot 3.4 beta 2 with the fixed Mono builds from https://github.com/godotengine/godot-mono-builds/pull/47

https://downloads.tuxfamily.org/godotengine/testing/Godot_v3.4-beta2_mono_with-btls_win64.zip

Could you confirm that it works as expected?

**Comment by sitiom** [comment]
Still gettting the error.
![image](https://user-images.githubusercontent.com/56180050/127635706-56adeba6-2de5-47bb-9a16-58733805f31b.png)

**Comment by akien-mga** [comment]
Thanks, something seems to still be off. Continuing debugging in -.

**Comment by norrath-hero** [comment]
https://github.com/godotengine/godot/issues/48258

**Comment by sitiom** [comment]
Note: Fixed in .NET 6 (https://github.com/godotengine/godot/pull/64089#issuecomment-1208887960)

**Comment by akien-mga** [comment]
This was closed again by mistake when updating a fork. But since this was fixed in 4.0, and no contributor is working on it for 3.x, I think we can leave it closed.
